tests/test_core.py

```diff
@@ -131,6 +147,215 @@ def test_filter_error(test_df):
+        unique_time = obs['time'].unique()
+        assert len(unique_time) == 1
+        assert unique_time[0] == expected
+
+
+@pytest.mark.parametrize("test_month", [6, "Jun", "jun", ["Jun", "jun"]])
+def test_filter_year_month(test_df, test_month):
+    if "year" in test_df.data.columns:
+        error_msg = re.escape("filter by `month` not supported")
+        with pytest.raises(ValueError, match=error_msg):
+            obs = test_df.filter(year=2005, month=test_month)
+    else:
+        obs = test_df.filter(year=2005, month=test_month)
+        expected = np.array(pd.to_datetime('2005-06-17T00:00:00.0'),
+                            dtype=np.datetime64)
+        unique_time = obs['time'].unique()
+        assert len(unique_time) == 1
+        assert unique_time[0] == expected
+
+
+@pytest.mark.parametrize("test_day", [17, "Fri", "Friday", "friday",
+                                      ["Fri", "fri"]])
+def test_filter_day(test_df, test_day):
+    if "year" in test_df.data.columns:
+        error_msg = re.escape("filter by `day` not supported")
+        with pytest.raises(ValueError, match=error_msg):
+            obs = test_df.filter(day=test_day)
+    else:
+        obs = test_df.filter(day=test_day)
+        expected = np.array(pd.to_datetime('2005-06-17T00:00:00.0'),
+                            dtype=np.datetime64)
+        unique_time = obs['time'].unique()
+        assert len(unique_time) == 1
+        assert unique_time[0] == expected
+
+
+@pytest.mark.parametrize("test_hour", [0, 12, [12, 13]])
+def test_filter_hour(test_df, test_hour):
+    if "year" in test_df.data.columns:
+        error_msg = re.escape("filter by `hour` not supported")
+        with pytest.raises(ValueError, match=error_msg):
+            obs = test_df.filter(hour=test_hour)
+    else:
+        obs = test_df.filter(hour=test_hour)
+        test_hour = [test_hour] if isinstance(test_hour, int) else test_hour
+        expected_rows = test_df.data["time"].apply(lambda x: x.hour)\
+            .isin(test_hour)
+        expected = test_df.data["time"].loc[expected_rows].unique()
+
+        unique_time = obs['time'].unique()
+        npt.assert_array_equal(unique_time, expected)
+
+
+def test_filter_time_exact_match(test_df):
+    if "year" in test_df.data.columns:
+        error_msg = "`year` can only be filtered with ints or lists of ints"
+        with pytest.raises(TypeError, match=re.escape(error_msg)):
+            test_df.filter(year=datetime.datetime(2005, 6, 17))
+    else:
+        obs = test_df.filter(time=datetime.datetime(2005, 6, 17))
+        expected = np.array(pd.to_datetime('2005-06-17T00:00:00.0'),
+                            dtype=np.datetime64)
+        unique_time = obs['time'].unique()
+        assert len(unique_time) == 1
+        assert unique_time[0] == expected
+
+
+def test_filter_time_range(test_df):
+    error_msg = re.escape(
+        "'datetime.datetime' object cannot be interpreted as an integer"
+    )
```

Comment:
this should be ```error_msg = r".*datetime.datetime.*"``` for Python2 compatibility, my bad

Reply:
thanks!